### PR TITLE
Add Ansible on WSL support

### PR DESCRIPTION
**Pull request by rmcolbert**
## Summary
Adds support for using Ansible installed inside WSL on Windows 10. 

To use WSL hosted Ansible, enable the "Use WSL" option in the extension preferences or add the following to settings.json:
    "ansible.useWSL": true

## Issue Type
- New Feature

### CODE REVIEW

**Comment by yungezz** [comment]
hi @zikalino  could you pls help to review the PR? thanks.

**Comment by rmcolbert** [comment]
Changed the implementation to use a WSL terminal instead of the default VSCode terminal the other methods use. This way of launching terminals may be useful in other scenarios like when running Node.

File: package.json [hunk 1]
```diff
@@ -230,6 +230,11 @@
           "default": "microsoft/ansible:latest",
           "description": "Custom docker image when running playbook in Docker."
         },
+        "ansible.useWSL": {
+          "type": "boolean",
+          "default": "",
+          "description": "Use local WSL to run Ansible"
```
Comment (yungezz) [inline package.json]: WSL to full name?

File: src/utilities.ts [hunk 2]
```diff
@@ -73,7 +74,7 @@ export function isDockerInstalled(outputChannel: vscode.OutputChannel, cb: Funct
 }
 
 export function isAnsibleInstalled(outputChannel: vscode.OutputChannel, cb: Function): void {
-    var cmd = process.platform === 'win32' ? 'ansible --version' : 'type ansible';
+    var cmd = process.platform === 'win32' ? 'wsl -- ansible --version' : 'type ansible';
```
Comment (yungezz) [inline src/utilities.ts]: if wsl is  not enabled?

File: src/utilities.ts [hunk 3]
```diff
@@ -114,6 +115,30 @@ export function IsNodeInstalled(): Promise<boolean> {
...
 
+export function IsWslInstalled(): Promise<boolean> {
+    var cmd = 'wsl.exe -- ls';
+    return new Promise<boolean>((resolve, reject) => {
+        child_process.exec(cmd).on('exit', function (code) {
+            if (!code) {
+                return resolve(true);
+            } else {
+                const open: vscode.MessageItem = { title: "View" };
+                vscode.window.showErrorMessage('Please install Windows Services for Linux.', open)
```
Comment (yungezz) [inline src/utilities.ts]: Windows Subsytem for Linux

File: src/localAnsibleRunner.ts [hunk 4]
```diff
@@ -31,7 +31,21 @@ export class LocalAnsibleRunner extends TerminalBaseRunner {
         if (utilities.isTelemetryEnabled()) {
             cmdsToTerminal.push(envCmd + Constants.UserAgentName + '=' + utilities.getUserAgent());
         }
-        cmdsToTerminal.push(this.getRunPlaybookCmd("\"" + playbook + "\""));
+
+        if (utilities.isWslEnabled()) {
+            var sourcePath = path.dirname(playbook);
+            var targetPath = '/playbook';
```
Comment (yungezz) [inline src/localAnsibleRunner.ts]: why the targetpath is hardcoded?

Comment (rmcolbert) [reply src/localAnsibleRunner.ts]: This was based on the existing path translation that is present in the dockerRunner.ts.

Comment (zikalino) [reply src/localAnsibleRunner.ts]: shouldn't it be sth like /mnt/c/….  in case of wsl?
just need to check how it works.

**Review by zikalino — CHANGES_REQUESTED** [review]
the pr looks very good and i am actually extremely happy to see it as i am using WSL personally :-)

let's just resolve the issue of target path and we can merge it!

File: src/localAnsibleRunner.ts (hunk 4 above)
Comment (yungezz) [reply src/localAnsibleRunner.ts]: For docker scenario, extension will mount local workspace as`playbook` in docker container. in wsl, it should be same folder name without any change, so not `playbook`.

Comment (rmcolbert) [reply src/localAnsibleRunner.ts]: I've clean up the playbook path translation

Comment (rmcolbert) [reply src/localAnsibleRunner.ts]: Just realized a usecase that needs to be addressed either in a future one PR ... playbook or workspace residing on a non-local drive or network path.

File: src/utilities.ts (hunk 2 above)
Comment (yungezz) [reply src/utilities.ts]: wsl is for windows 10 and later. there's option to run ansible on cygwin on windows.

File: src/localAnsibleRunner.ts [hunk 5]
```diff
@@ -31,7 +31,21 @@ export class LocalAnsibleRunner extends TerminalBaseRunner {
         if (utilities.isTelemetryEnabled()) {
             cmdsToTerminal.push(envCmd + Constants.UserAgentName + '=' + utilities.getUserAgent());
         }
-        cmdsToTerminal.push(this.getRunPlaybookCmd("\"" + playbook + "\""));
+
+        if (utilities.isWslEnabled()) {
+            var sourcePath = path.dirname(playbook);
+            var targetPath = '/mnt/' + playbook.replace(/:/, '');
```
Comment (yungezz) [inline src/localAnsibleRunner.ts]: `:` to `/`, not empty

**Comment by yungezz** [comment]
how do you do the testing?

File: src/localAnsibleRunner.ts (hunk 5 above)
Comment (rmcolbert) [reply src/localAnsibleRunner.ts]: I have to disagree. Absolute paths in windows are <drive>:\path. You don't want to replace : with / or you would end up with /mnt/c//something.

Comment (yungezz) [reply src/localAnsibleRunner.ts]: got it.

Comment (rmcolbert) [reply src/localAnsibleRunner.ts]: > how do you do the testing?

I test on Windows 10 Pro 1903 compiling the extension from source and validating against WSL running Ubuntu 18.04 testing w/ using the WSL instance as both a task engine and local.

**Review by yungezz — APPROVED** [review]

**Comment by yungezz** [comment]
@zikalino  could you pls check the PR again?

File: src/utilities.ts (hunk 2 above)
Comment (rmcolbert) [reply src/utilities.ts]: While I agree that cygwin does make it possible to run Ansible on Windows, instructions for installing Ansible on WSL is actually documented on the Ansible docs site. https://docs.ansible.com/ansible/latest/user_guide/windows_faq.html#can-ansible-run-on-windows

If this is really wanted & used by the community, I'll gladly put that check back in.

**Review by zikalino — APPROVED** [review]

**Comment by zikalino** [comment]
cool so now we can merge :-)
I will test myself tomorrow. today too busy!
thank you @rmcolbert !

**Comment by andrew-from-toronto** [comment]
Any update on this? I would love to be able to run my ansible playbooks (in wsl) right within vscode when I'm working in Windows

**Comment by rmcolbert** [comment]
So this original feature is still useful for some but since Microsoft created the "Remote - WSL" extension, you can open your Ansible folder in VSCode connected to your WSL instance and deploy the Ansible extension inside the WSL instance and use the "Run Playbook in Local Ansible" option.

**Comment by andrew-from-toronto** [comment]
Very cool, thanks for explaining that @rmcolbert !

**Pull request merged**